tools/wake-format/emitter.h

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2022 SiFive, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You should have received a copy of LICENSE.Apache2 along with
+ * this software. If not, you may obtain a copy at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <wcl/optional.h>
+#include <wcl/rope.h>
+
+#include <cassert>
+#include <iostream>
+#include <vector>
+
+#include "parser/cst.h"
+
+class Emitter {
+ public:
+  struct ctx_t {
+    int nest_level = 0;
+    bool is_flat = false;
+
+    ctx_t flat() {
+      ctx_t copy = *this;
+      copy.is_flat = true;
+      return copy;
+    }
+  };
+
+  struct nest_t {
+    ctx_t* ctx;
+
+    nest_t(ctx_t* ctx) : ctx(ctx) { ctx->nest_level++; }
+
+    ~nest_t() { ctx->nest_level--; }
+  };
+
+  Emitter(std::ostream* ostream) : ostream(ostream) { assert(ostream != nullptr); }
```

Comment:
Now that we're returning ropes, can we just avoid the ostream? Like we compute the correct rope and then that rope can write to the output stream later. We always have a place to write to and we memoize so I think it decouples ostream creation from the emitter. The thing I didn't like previously was something (I've forgotten what now) that you were forced to do because of needing the ostream.